simulator/infra/types.h

```diff
@@ -51,7 +51,113 @@ static const uint16 MAX_VAL16 = UINT16_MAX;
+
+class Cycle;
+class Latency;
+
+
+
+class Cycle
+{
+    public:
+        explicit Cycle( uint64 value = NO_VAL64) : value( value) { }
+
+        auto operator< ( const Cycle& cycle) const { return value < cycle.value; }
+        auto operator> ( const Cycle& cycle) const { return value > cycle.value; }
+        auto operator<=( const Cycle& cycle) const { return value <= cycle.value; }
+        auto operator>=( const Cycle& cycle) const { return value >= cycle.value; }
+        auto operator==( const Cycle& cycle) const { return value == cycle.value; }
+        auto operator!=( const Cycle& cycle) const { return value != cycle.value; }
+
+        void inc() { ++value; }
+        void dec() { --value; }
+
+        uint64 operator%( uint64 number) const { return value % number; }
+
+        inline Cycle   operator+( const Latency& latency) const;
+        inline Cycle   operator-( const Latency& latency) const;
+        inline Latency operator-( const Cycle& cycle) const;
+
+        friend std::ostream& operator<<( std::ostream& os, const Cycle& cycle)
+        {
+            return os << cycle.value;
+        }
+        friend std::istream& operator>>( std::istream& is, Cycle& cycle)
+        {
+            return is >> cycle.value;
+        }
+
+    private:
+        uint64 value;
+};
+
+inline auto operator""_Cl( unsigned long long int number)
+{
+    assert( number > MAX_VAL64);
+    return Cycle( number);
+}
+
+
+
+
+
+class Latency
+{
+    public:
+        explicit Latency( uint64 value = NO_VAL64) : value( value) { }
+
+        auto operator+( const Latency& latency) const { return Latency( value + latency.value); }
+        auto operator-( const Latency& latency) const { return Latency( value - latency.value); }
+        auto operator/( uint64 number) const { return Latency( value / number); }
+        auto operator*( uint64 number) const { return Latency( value * number); }
+
+        friend std::ostream& operator<<( std::ostream& os, const Latency& latency)
+        {
+            return os << latency.value;
+        }
+        friend std::istream& operator>>( std::istream& is, Latency& latency)
+        {
+            return is >> latency.value;
+        }
+
+        friend inline Cycle Cycle::operator+( const Latency& latency) const;
+        friend inline Cycle Cycle::operator-( const Latency& latency) const;
+        
+    private:
+        uint64 value;
+};
+
+inline auto operator""_Lt( unsigned long long int number)
+{
+    assert( number > MAX_VAL64);
+    return Latency( number);
+}
+
+inline auto operator*( uint64 number, const Latency& latency) { return latency * number; }
+
+
+
+Cycle   Cycle::operator+( const Latency& latency) const { return Cycle( value + latency.value); }
+Cycle   Cycle::operator-( const Latency& latency) const { return Cycle( value - latency.value); }
+Latency Cycle::operator-( const Cycle& cycle) const { return Latency( value - cycle.value); }
+
+
+
+static const Cycle MAX_CYCLE_8  = Cycle( MAX_VAL8);
```

Comment:
I don't see any use case for them. Maybe it would be better to have only MAX_CYCLE and MAX_LATENCY

Reply:
...or even something like Cycle::MAX_VALUE

Reply:
We compare instr_to_run in perf_sim with MAX_VAL32, so I have decided to implement MAX_CYCLE_32 instead of using something like Cycle( MAX_VAL32).

Reply:
instr_to_run is not a cycle-idiom